### PR TITLE
Move score script context from SearchScript to its own class

### PR DESCRIPTION
**Pull request by martijnvg**

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreFunction.java [hunk 1]
```diff
@@ -58,27 +58,26 @@ public DocIdSetIterator iterator() {
...
         this.sScript = sScript;
         this.script = script;
     }
 
     @Override
     public LeafScoreFunction getLeafScoreFunction(LeafReaderContext ctx) throws IOException {
-        final SearchScript leafScript = script.newInstance(ctx);
+        final ScoreScript leafScript = script.newInstance(ctx);
         final CannedScorer scorer = new CannedScorer();
-        leafScript.setScorer(scorer);
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreFunction.java]: I think we want to keep the scorer. Having the script access score through `get_score()` (which maps to `_score`) allows score to not be computed when it is not used. needs_score only gets us so far (whether it is used at all). With get_score, we could (although I'm not sure it is actually implemented this way yet) call the method lazily to bind _score.

**Review by rjernst — COMMENTED** [review]
This is fantastic! I just have one comment.

**Comment by martijnvg** [comment]
Thanks for reviewing! I've updated the pr and brought back setScorer() and added get_score() which is backed by a supplier that lazily computes the score if needed.

**Review by rjernst — APPROVED** [review]
LGTM once the CI issue is addressed (need to add new ScoreScript to MockScriptEngine)

**Pull request merged**